### PR TITLE
Refs #14684 - Triggering validation

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Comment by ares** [comment]
tests failed

File: app/models/foreman_tasks/triggering.rb [hunk 1]
```diff
@@ -39,17 +35,27 @@ class Triggering < ActiveRecord::Base
...
+        triggering.recurring_logic = ::ForemanTasks::RecurringLogic.new_from_triggering(triggering) if triggering.recurring?
       end
     end
 
+    def mode
+      read_attribute(:mode).to_sym
+    end
+
+    def mode=(mod)
+      write_attribute :mode, mod.downcase.to_sym
```
Comment (ares) [inline app/models/foreman_tasks/triggering.rb]: `mod` is user input, we should make sure it's white-listed in ALLOWED_MODES.map(&:to_s) before converting it to symbol

File: app/models/foreman_tasks/triggering.rb [hunk 2]
```diff
@@ -93,10 +98,16 @@ def parse_start_before!
 
     private
 
-    def correct_cronline
-      ForemanTasks::RecurringLogic.new_from_cronline(cronline).next_occurrence_time
-    rescue ArgumentError => _
-      self.errors.add(:cronline, _("%s is not valid format of cron line")) % (cronline)
+    def can_start_recurring
+      recurring_logic.next_occurrence_time
```
Comment (ares) [inline app/models/foreman_tasks/triggering.rb]: this line doesn't seem to be necessary, `valid?` nor `valid_cronline?` seems to rely on @parser being set, `valid_cronline?` calls it on its own

File: app/models/foreman_tasks/triggering.rb [hunk 3]
```diff
@@ -39,17 +35,27 @@ class Triggering < ActiveRecord::Base
...
         triggering.mode = params.fetch(:mode, :immediate).to_sym
         triggering.input_type = params.fetch(:input_type, :daily).to_sym
         triggering.end_time_limited = params[:end_time_limited] == "true"
         triggering.start_at_raw ||= Time.now.strftime(TIME_FORMAT)
+        triggering.recurring_logic = ::ForemanTasks::RecurringLogic.new_from_triggering(triggering) if triggering.recurring?
       end
     end
 
+    def mode
+      read_attribute(:mode).to_sym
```
Comment (ares) [inline app/models/foreman_tasks/triggering.rb]: `super.to_sym` might work

also you could use your own serializer (only if you prefer, I don't have strong opinion) - http://apidock.com/rails/ActiveRecord/AttributeMethods/Serialization/ClassMethods/serialize#1343-Custom-serialization

File: app/models/foreman_tasks/triggering.rb (hunk 1 above)
Comment (ares) [inline app/models/foreman_tasks/triggering.rb]: `super(mod.downcase.to_sym)` might work after `mod` is validated

**Comment by ares** [comment]
It fixes the immediate  invocation issue :+1:
for delayed one I don't see validation error, it now creates invocation which remains queued until I manually cancel the recurring logic, can we somehow either don't allow this (that's actually what I've expected after reading the code) or make the recurring logic Finished?

**Comment by adamruzicka** [comment]
@ares have you tried it together with https://github.com/theforeman/foreman_remote_execution/pull/172 ? The REX PR is required to actually use the validations added to the triggering in this PR

**Comment by ares** [comment]
oops I meanwhile reset the branch there, sorry let me retest

**Comment by ares** [comment]
works fine now, so last two things to consider are accessors methods

**Comment by iNecas** [comment]
[test]

**Comment by ares** [comment]
ACK from my side

**Comment by iNecas** [comment]
Squashed and merged as https://github.com/theforeman/foreman-tasks/commit/a637787a5d3963e2a16e57e905dc31f19ce07680 Thanks @adamruzicka

**Pull request closed without merging**